cpp/arcticdb/storage/azure/azure_real_client.cpp

```diff
@@ -0,0 +1,115 @@
+    }
+    return client_options;
+}
+
+void RealAzureClient::write_blob(
+        const std::string& blob_name,
+        Segment&& segment,
+        const Azure::Storage::Blobs::UploadBlockBlobFromOptions& upload_option,
+        unsigned int request_timeout) {
+
+    std::shared_ptr<Buffer> tmp;
+    auto hdr_size = segment.segment_header_bytes_size();
+    auto [dst, write_size] = segment.try_internal_write(tmp, hdr_size);
+    util::check(arcticdb::Segment::FIXED_HEADER_SIZE + hdr_size + segment.buffer().bytes() <= write_size,
+                "Size disparity, fixed header size {} + variable header size {} + buffer size {}  >= total size {}",
+                arcticdb::Segment::FIXED_HEADER_SIZE,
+                hdr_size,
+                segment.buffer().bytes(),
+                write_size);
+    ARCTICDB_SUBSAMPLE(AzureStorageUploadObject, 0)
+    auto blob_client = container_client.GetBlockBlobClient(blob_name);
+    ARCTICDB_RUNTIME_DEBUG(log::storage(), "Writing key '{}' with {} bytes of data",
+                           blob_name,
+                           segment.total_segment_size(hdr_size));
+    blob_client.UploadFrom(dst, write_size, upload_option, get_context(request_timeout));
+    ARCTICDB_RUNTIME_DEBUG(log::storage(), "Wrote key '{}' with {} bytes of data",
+                           blob_name,
+                           segment.total_segment_size(hdr_size));
+}
+
+Segment RealAzureClient::read_blob(
+        const std::string& blob_name,
+        const Azure::Storage::Blobs::DownloadBlobToOptions& download_option,
+        unsigned int request_timeout) {
+
+    ARCTICDB_DEBUG(log::storage(), "Looking for blob {}", blob_name);
+    auto blob_client = container_client.GetBlockBlobClient(blob_name);
+    auto properties = blob_client.GetProperties(Azure::Storage::Blobs::GetBlobPropertiesOptions{}, get_context(request_timeout)).Value;
+    std::shared_ptr<Buffer> buffer = std::make_shared<Buffer>(properties.BlobSize);
+    blob_client.DownloadTo(buffer->preamble(), buffer->available(), download_option, get_context(request_timeout));
+    ARCTICDB_SUBSAMPLE(AzureStorageVisitSegment, 0)
+
+    return Segment::from_buffer(std::move(buffer));
+}
+
+void RealAzureClient::delete_blobs(
+        const std::vector<std::string>& blob_names,
+        unsigned int request_timeout) {
+    auto batch = container_client.CreateBatch();
+
+    for (auto& blob_name: blob_names) {
+        ARCTICDB_RUNTIME_DEBUG(log::storage(), "Removing azure blob with key {}", blob_names);
+        batch.DeleteBlob(blob_name);
+    }
+
+    ARCTICDB_RUNTIME_DEBUG(log::storage(), "Submitting DeleteBlob batch");
+    ARCTICDB_SUBSAMPLE(AzureStorageDeleteObjects, 0)
+    container_client.SubmitBatch(batch, Azure::Storage::Blobs::SubmitBlobBatchOptions(), get_context(request_timeout));//To align with s3 behaviour, deleting non-exist objects is not an error, so not handling response
```

Comment:
Haven't we lost the batching logic here? master has logic to submit deletes in batches of size no greater than `delete_object_limit`? Without the batching, deletes for blobs outside of the batch size get silently dropped (effectively breaking deletion).

Reply:
Ah I see OK, so you've just lifted the batching logic out and this method assumes that `blob_names` is within the batch size, which makes sense :+1: I think it would be worth adding an assertion that `blob_names` is within the max batch size?

Reply:
Done in [b9a6e99](https://github.com/man-group/ArcticDB/pull/1315/commits/b9a6e99b7738e1623c1ed4af80411744a86b0fa6).